settings: Display translated message after updating default language.

haocheng6 opened the pull request:
This commit updates the /json/settings endpoint to return a message in the preferred language that tells the user to reload the browser and updates the UI to display the message when the default language has been changed.

The translated message is returned as the `msg` property of the response data, but I am not sure that is the intended usage of the `msg` property and whether it is more desirable to add a new field to the response.

The test in `settings.test.ts` relies on the Simplified Chinese translation of the helper message. The test is currently commented out because we don't have the necessary translation in `locale/zh_Hans/LC_MESSAGES/django.po`.

<!-- Describe your pull request here.-->

Fixes: #25552

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**

(Note: to view the effect of this PR, the `django.po` file for different languages should have a translation for the sentence "Saved. Please <a class='reload_link'>reload</a> for the change to take effect." I manually added one to `locale/zh_Hans/LC_MESSAGES/django.po` before taking the screenshot below.)

The screenshot shows the message displayed after the default language is changed from English to Simplified Chinese.

![translated_reload_message](https://github.com/zulip/zulip/assets/7645930/bde12576-173e-48aa-872f-7bbee1d7122f)

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

alya commented:
@sahil839 could you please take a look at this one?

sahil839 commented on web/src/settings_ui.ts:
```diff
@@ -55,6 +55,16 @@ export function do_settings_change(
+                response_data.msg
+            ) {
+                success_msg_html = response_data.msg;
+            }
```
Instead of modifying this function which is used by many settings, would it be better to just handle default language setting as a special case to show the success response from server since for all other settings we do not have any success response from server.

haocheng6 replied:
Thanks for the feedback and sorry for the delay.

I think the `do_settings_change` function has to be changed in some way because that is the only place where `response_data` is available as far as I know, but I agree it would be better to treat default language as a special case.

In the new commit (4488e78f1ccd8d34454a248a817d53b90195a50f), I added a new option `success_msg_html_generator` to the `do_settings_change` function to generate a success message from the response. Currently, `success_msg_html_generator` is only used when `default_language` is part of the settings to be updated. Is this okay?

sahil839 replied:
I was thinking of directly using `channel.patch` instead of using `do_settings_change` or `change_display_setting`.

haocheng6 replied:
Thanks for the clarification! I reverted changes made to `do_settings_change` and `change_display_setting` and added a new function to handle default language setting as a special case (see 99c07f075567bda665209615bdfac7fb26f762b6 and ade85f9659012eb4cce8757de54d6f3e0d6d339a).

haocheng6 commented on zerver/views/user_settings.py:
```diff
@@ -376,6 +377,11 @@ def json_change_settings(
+            if k == "default_language":
+                with override_language(v):
+                    result["msg"] = _(
+                        "Saved. Please <a class='reload_link'>reload</a> for the change to take effect."
```
Would it be better to add a new field like `success_msg_html` to the response when updating the default language?

sahil839 replied:
I am not sure about this. @timabbott thoughts?

sahil839 commented on web/src/settings_display.js:
```diff
@@ -103,11 +100,11 @@ function user_default_language_modal_post_render() {
             );
 
-            change_display_setting(
+            settings_ui.do_default_language_setting_change(
```
There is no need for a separate function here. You can directly call channel.patch.

Also, please fix the commit history here and remove unnecessary commits. See https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html for details.

haocheng6 replied:
Done here: 2d756845f5d2286b7be5fb8ee0cc36013d83603d.

haocheng6 committed:
settings: Display translated message after updating default language.

This commit updates the /json/settings endpoint to return a message
in the preferred language that tells the user to reload the browser
and updates the UI to display the message when the default language
has been changed.

Fixes #25552.

sahil839 commented on web/src/settings_display.js:
```diff
@@ -100,21 +100,25 @@ function user_default_language_modal_post_render() {
+                success(response_data) {
+                    const appear_after = 500;
+                    const success_msg_html =
+                        response_data.msg || $t_html({defaultMessage: "Saved"});
```
Is there any need to add  the "||" condition here? We would always expect a message returned by server.

shuklamaneesh23 commented:
Hey @haocheng6, will you be continuing this work, or should I take it to completion?
Thanks!
